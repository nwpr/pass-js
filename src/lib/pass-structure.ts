/* eslint-disable max-depth */
/**
 * Class that implements base structure fields setters / getters
 *
 * @see {@link https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3}
 */

import { PASS_STYLES, STRUCTURE_FIELDS, TRANSIT } from '../constants.js';
import {
  ApplePass,
  BoardingPass,
  NFCDictionary,
  PassCommonStructure,
  PassStyle,
  TransitType,
} from '../interfaces.js';

import { FieldsMap } from './fieldsMap.js';
import { NFCField } from './nfc-fields.js';

export class PassStructure {
  [key: string]: unknown;
  protected fields: Partial<ApplePass> = {};

  constructor(fields: Partial<ApplePass> = {}) {
    // setting style first
    for (const style of PASS_STYLES) {
      if (style in fields) {
        this.style = style;
        if ('boardingPass' in fields && fields.boardingPass) {
          this.transitType = (fields.boardingPass as BoardingPass)
            .transitType as TransitType;
        } else if ('storeCard' in this.fields && 'nfc' in fields) {
          // check NFC fields
          this.fields.nfc = new NFCField(fields.nfc as NFCDictionary);
        }
        const structure: PassCommonStructure = fields[
          this.style
        ] as PassCommonStructure;
        for (const prop of STRUCTURE_FIELDS) {
          if (prop in structure) {
            const currentProperty = structure[prop];
            if (Array.isArray(currentProperty))
              for (const field of currentProperty)
                (this[prop] as FieldsMap).add(field);
            else if (currentProperty instanceof FieldsMap)
              // copy fields
              for (const [key, data] of currentProperty)
                (this[prop] as FieldsMap).add({ key, ...data });
          }
        }
      }
    }
  }

  /**
   * Pass type, e.g boardingPass, coupon, etc
   */
  get style(): PassStyle | undefined {
    for (const style of PASS_STYLES) {
      if (style in this.fields) return style;
    }
    return undefined;
  }

  /**
   * Required for boarding passes; otherwise not allowed.
   * Type of transit.
   * Must be one of the following values: PKTransitTypeAir, PKTransitTypeBoat, PKTransitTypeBus, PKTransitTypeGeneric,PKTransitTypeTrain.
   */
  get transitType(): TransitType | undefined {
    if (this.style !== 'boardingPass')
      throw new ReferenceError(
        `transitType field only allowed in Boarding Passes, current pass is ${this.style}`,
      );
    if ('boardingPass' in this.fields && this.fields.boardingPass)
      return (this.fields.boardingPass as BoardingPass).transitType as
        | TransitType
        | undefined;
    return undefined;
  }

  /**
   * NFC-enabled pass keys support sending reward card information as part of an Apple Pay transaction.
   *
   * NFC-enabled pass keys are only supported in passes that contain an Enhanced Passbook/NFC certificate.
   * For more information, contact merchant support at https://developer.apple.com/contact/passkit/.
   * **Only for storeCards with special Apple approval**
   *
   * @see {@link https://developer.apple.com/library/archive/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/TopLevel.html#//apple_ref/doc/uid/TP40012026-CH2-DontLinkElementID_3}
   */
  get nfc(): NFCField {
    if (!('storeCard' in this.fields))
      throw new ReferenceError(
        `NFC fields only available for storeCard passes, current is ${this.style}`,
      );
    return this.fields.nfc as NFCField;
  }

  get headerFields(): FieldsMap {
    const { style } = this;
    if (!style)
      throw new ReferenceError(
        `Pass style is undefined, set the pass style before accessing pass structure fields`,
      );
    const styleFields: PassCommonStructure = this.fields[
      style
    ] as PassCommonStructure;
    if (!(styleFields.headerFields instanceof FieldsMap))
      styleFields.headerFields = new FieldsMap();
    return styleFields.headerFields;
  }
  get auxiliaryFields(): FieldsMap {
    const { style } = this;
    if (!style)
      throw new ReferenceError(
        `Pass style is undefined, set the pass style before accessing pass structure fields`,
      );
    const styleFields: PassCommonStructure = this.fields[
      style
    ] as PassCommonStructure;
    if (!(styleFields.auxiliaryFields instanceof FieldsMap))
      styleFields.auxiliaryFields = new FieldsMap();
    return styleFields.auxiliaryFields;
  }
  get backFields(): FieldsMap {
    const { style } = this;
    if (!style)
      throw new ReferenceError(
        `Pass style is undefined, set the pass style before accessing pass structure fields`,
      );
    const styleFields: PassCommonStructure = this.fields[
      style
    ] as PassCommonStructure;
    if (!(styleFields.backFields instanceof FieldsMap))
      styleFields.backFields = new FieldsMap();
    return styleFields.backFields;
  }
  get primaryFields(): FieldsMap {
    const { style } = this;
    if (!style)
      throw new ReferenceError(
        `Pass style is undefined, set the pass style before accessing pass structure fields`,
      );
    const styleFields: PassCommonStructure = this.fields[
      style
    ] as PassCommonStructure;
    if (!(styleFields.primaryFields instanceof FieldsMap))
      styleFields.primaryFields = new FieldsMap();
    return styleFields.primaryFields;
  }
  get secondaryFields(): FieldsMap {
    const { style } = this;
    if (!style)
      throw new ReferenceError(
        `Pass style is undefined, set the pass style before accessing pass structure fields`,
      );
    const styleFields: PassCommonStructure = this.fields[
      style
    ] as PassCommonStructure;
    if (!(styleFields.secondaryFields instanceof FieldsMap))
      styleFields.secondaryFields = new FieldsMap();
    return styleFields.secondaryFields;
  }

  set style(v: PassStyle | undefined) {
    // remove all other styles
    for (const style of PASS_STYLES) if (style !== v) delete this.fields[style];
    if (!v) return;
    if (!PASS_STYLES.has(v)) throw new TypeError(`Invalid Pass type "${v}"`);
    if (!(v in this.fields)) this.fields[v] = {};
    // Add NFC fields
    if ('storeCard' in this.fields) this.fields.nfc = new NFCField();
    //   if ('boardingPass' in this.fields && this.fields.boardingPass) this.fields.boardingPass.
  }

  set transitType(v: TransitType | undefined) {
    const { style } = this;
    if (!style) {
      // removing transitType on empty pass does nothing
      if (!v) return;
      // setting transitStyle on a pass without type will set this pass as boardingPass also
      this.style = 'boardingPass';
    }
    if (!('boardingPass' in this.fields))
      throw new ReferenceError(
        `transitType field is only allowed at boarding passes`,
      );

    if (!v) {
      if (this.fields.boardingPass)
        delete (this.fields.boardingPass as BoardingPass).transitType;
    } else {
      if (Object.values(TRANSIT).includes(v)) {
        if (this.fields.boardingPass)
          (this.fields.boardingPass as BoardingPass).transitType = v;
        else this.fields.boardingPass = { transitType: v };
      } else throw new TypeError(`Unknown transit type "${v}"`);
    }
  }
}
