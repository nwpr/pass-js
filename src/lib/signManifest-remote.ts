import fetch from 'node-fetch';

const URL = process.env.URL || 'https://localhost:3000/api/sign-manifest';
const HEADERS = process.env.HEADERS || {};
HEADERS['Content-Type'] = 'application/json';

/**
 * Signs a manifest and returns the signature.
 *
 * @param {string} manifest - manifest to sign
 * @returns {Buffer} - signature for given manifest
 */
export async function signManifest(manifest: string): Promise<Buffer> {
  return await fetch(URL, {
    method: 'POST',
    headers: HEADERS,
    body: manifest,
  })
    .then(async (res) => {
      if (res.status !== 200) {
        throw new Error(`Failed to sign manifest: ${res.statusText}`);
      }
      return res.buffer();
    })
    .catch(err => {
      throw new Error(`Failed to sign manifest: ${err}`);
    });
}
