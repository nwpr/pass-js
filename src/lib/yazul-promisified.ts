import { promisify } from 'util';

import { EventIterator } from 'event-iterator';
import { Entry, Options, ZipFile, fromBuffer as ZipFromBuffer } from 'yauzl';
import { Readable } from 'stream';

import { streamToBuffer } from './stream-to-buffer.js';

// Promisifying yauzl
Object.defineProperties(ZipFile.prototype, {
  [Symbol.asyncIterator]: {
    enumerable: true,
    writable: false,
    configurable: false,
    value() {
      return new EventIterator<Entry>(queue => {
        this.addListener('entry', queue.push);
        this.addListener('end', queue.stop);
        this.addListener('error', queue.fail);
      })[Symbol.asyncIterator]();
    },
  },
  openReadStreamAsync: {
    enumerable: true,
    writable: false,
    configurable: false,
    // eslint-disable-next-line @typescript-eslint/unbound-method
    value: promisify(ZipFile.prototype.openReadStream),
  },
  getBuffer: {
    enumerable: true,
    writable: false,
    configurable: false,
    async value(entry: Entry) {
      const stream: Readable = await this.openReadStreamAsync(entry);
      return streamToBuffer(stream);
    },
  },
});
export const unzipBuffer = promisify(ZipFromBuffer) as unknown as (
  buffer: Buffer,
  options?: Options,
) => Promise<
  ZipFile & {
    openReadStreamAsync: (v: Entry) => Promise<import('stream').Readable>;
    getBuffer: (entry: Entry) => Promise<Buffer>;
    [Symbol.asyncIterator](): AsyncIterator<Entry>;
  }
>;
