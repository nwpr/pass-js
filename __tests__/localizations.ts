/* eslint-disable no-shadow */
import * as path from 'path';
import { TextDecoder } from 'util';
import { execFileSync } from 'child_process';
import { randomBytes } from 'crypto';
import { unlinkSync, mkdtempSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { toEndWith, toIncludeSameMembers } from 'jest-extended';
expect.extend({ toEndWith, toIncludeSameMembers });

import * as Localization from '../src/lib/localizations';

const RU_STRINGS_FILE = path.resolve(
  __dirname,
  './resources/passes/Generic/ru.lproj/pass.strings',
);
const ZH_STRINGS_FILE = path.resolve(
  __dirname,
  './resources/passes/Generic/zh_CN.lproj/pass.strings',
);

describe('Localizations files helpers', () => {
  it('escapeString -> unescape', () => {
    const str = `This is "multiline" string',
      with some rare character like \\ and \\\\ üòú inside it`;
    expect(Localization.unescapeString(Localization.escapeString(str))).toBe(
      str,
    );
  });

  if (process.platform === 'darwin') {
    it('string fixtures files must pass plutil -lint', async () => {
      const stdout = execFileSync(
        'plutil',
        ['-lint', RU_STRINGS_FILE, ZH_STRINGS_FILE],
        {
          encoding: 'utf8',
        },
      );
      expect(stdout.trim().split(/\n/)).toEqual(
        expect.arrayContaining([expect.toEndWith(': OK')]),
      );
    });
  }
  it('should read pass.strings file', async () => {
    const resRu = await Localization.readLprojStrings(RU_STRINGS_FILE);
    expect([...resRu]).toMatchSnapshot();
    const resZh = await Localization.readLprojStrings(ZH_STRINGS_FILE);
    expect([...resZh]).toMatchSnapshot();
  });

  it('returns string as utf-16 .lproj buffer', async () => {
    const resRu = await Localization.readLprojStrings(RU_STRINGS_FILE);
    const buf = Localization.getLprojBuffer(resRu);
    expect(Buffer.isBuffer(buf)).toBeTruthy();
    const decoder = new TextDecoder('utf-16', { fatal: true });
    expect(decoder.decode(buf)).toMatchSnapshot();
  });

  if (process.platform === 'darwin') {
    it('output buffer passes plutil -lint', async () => {
      const resZh = await Localization.readLprojStrings(ZH_STRINGS_FILE);
      const tmd = mkdtempSync(`${tmpdir()}${path.sep}`);
      const stringsFileName = path.join(
        tmd,
        `pass-${randomBytes(10).toString('hex')}.strings`,
      );
      writeFileSync(stringsFileName, Localization.getLprojBuffer(resZh));

      const stdout = execFileSync('plutil', ['-lint', stringsFileName], {
        encoding: 'utf8',
      });
      unlinkSync(stringsFileName);
      expect(stdout.trim()).toEndWith(': OK');
    });
  }

  it('loads localizations from files', async () => {
    const loc = new Localization.Localizations();
    await loc.load(path.resolve(__dirname, './resources/passes/Generic'));
    expect(loc.size).toBe(2);
    // ensure it normalizes locale
    expect(loc.has('zh-CN')).toBeTruthy();
  });

  it('read -> write -> compare', async () => {
    const resRu = await Localization.readLprojStrings(RU_STRINGS_FILE);
    const tmd = mkdtempSync(`${tmpdir()}${path.sep}`);
    const stringsFileName = path.join(
      tmd,
      `pass-${randomBytes(10).toString('hex')}.strings`,
    );
    writeFileSync(stringsFileName, Localization.getLprojBuffer(resRu));

    const resRu2 = await Localization.readLprojStrings(stringsFileName);

    // If we on macOs we can't miss an opportunity to plutil -lint
    if (process.platform === 'darwin') {
      const stdout = execFileSync('plutil', ['-lint', stringsFileName], {
        encoding: 'utf8',
      });
      expect(stdout.trim()).toEndWith(': OK');
    }
    unlinkSync(stringsFileName);
    expect([...resRu]).toIncludeSameMembers([...resRu2]);
  });

  it('should clone other instance if provided for constructor', () => {
    const loc1 = new Localization.Localizations();
    loc1
      .add('ru', { key1: 'test key 1', key2: 'test key 2' })
      .add('fr', { key1: 'test fr key1', key2: 'test fr key2' });
    const loc2 = new Localization.Localizations(loc1);
    expect(loc2.size).toBe(2);
    expect(loc2.get('fr').get('key2')).toBe('test fr key2');
    // modify a key in original loc1
    loc1.get('ru').set('key1', '—Ç–µ—Å—Ç');
    expect(loc2.get('ru').get('key1')).toBe('test key 1');
    expect(loc1.toArray()).toMatchSnapshot();
  });
});
