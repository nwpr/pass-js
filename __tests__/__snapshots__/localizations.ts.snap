// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Localizations files helpers returns string as utf-16 .lproj buffer 1`] = `
""Terms of Service" = "Условия использования";
"Privacy Policy" = "Политика конфиденциальности";
"Legal Notices" = "Юридические уведомления";
"Unable to load webpage." = "Не удалось загрузить веб-страницу.";
"OK" = "ОК";
"Do you want to add the token named “%@”?" = "Добавить токен \\"%@\\"?";
"No" = "Нет";
"Yes" = "Да";
"Invalid Key" = "Недействительный ключ";
"The key '%@' is invalid." = "Ключ \\"%@\\" недействителен.";
"You must enter a key." = "Необходимо ввести ключ.";
"Try Again" = "Повторить попытку";
"Cancel" = "Отмена";
"Invalid Barcode" = "Недопустимый штрихкод";
"The barcode '%@' is not a valid authentication token barcode." = "Штрихкод \\"%@\\" не является действительным токеном аутентификации.";
"Welcome" = "Добро пожаловать";
"Welcome_label" = "Google Authenticator обеспечивает дополнительную безопасность аккаунта с помощью сгенерированных кодов.\\n\\nЧтобы использовать Google Authenticator, необходимо добавить аккаунт, нажав:";
"Legal Information" = "Юридическая информация";
"Authenticator" = "Authenticator";
"Google Authenticator" = "Google Authenticator";
"Add Token" = "Добавить токен";
"Time Based" = "По времени";
"Counter Based" = "По счетчику";
"Account:" = "Аккаунт:";
"user@example.com" = "user@example.com";
"Key:" = "Ключ:";
"Enter your key" = "Введите ключ";
"Scan Barcode" = "Сканировать штрихкод";
"Integrity Check Value" = "Integrity Check Value";"
`;

exports[`Localizations files helpers should clone other instance if provided for constructor 1`] = `
[
  {
    "data": {
      "data": [
        255,
        254,
        34,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        49,
        0,
        34,
        0,
        32,
        0,
        61,
        0,
        32,
        0,
        34,
        0,
        66,
        4,
        53,
        4,
        65,
        4,
        66,
        4,
        34,
        0,
        59,
        0,
        10,
        0,
        34,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        50,
        0,
        34,
        0,
        32,
        0,
        61,
        0,
        32,
        0,
        34,
        0,
        116,
        0,
        101,
        0,
        115,
        0,
        116,
        0,
        32,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        32,
        0,
        50,
        0,
        34,
        0,
        59,
        0,
      ],
      "type": "Buffer",
    },
    "path": "ru.lproj/pass.strings",
  },
  {
    "data": {
      "data": [
        255,
        254,
        34,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        49,
        0,
        34,
        0,
        32,
        0,
        61,
        0,
        32,
        0,
        34,
        0,
        116,
        0,
        101,
        0,
        115,
        0,
        116,
        0,
        32,
        0,
        102,
        0,
        114,
        0,
        32,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        49,
        0,
        34,
        0,
        59,
        0,
        10,
        0,
        34,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        50,
        0,
        34,
        0,
        32,
        0,
        61,
        0,
        32,
        0,
        34,
        0,
        116,
        0,
        101,
        0,
        115,
        0,
        116,
        0,
        32,
        0,
        102,
        0,
        114,
        0,
        32,
        0,
        107,
        0,
        101,
        0,
        121,
        0,
        50,
        0,
        34,
        0,
        59,
        0,
      ],
      "type": "Buffer",
    },
    "path": "fr.lproj/pass.strings",
  },
]
`;

exports[`Localizations files helpers should read pass.strings file 1`] = `
[
  [
    "Terms of Service",
    "Условия использования",
  ],
  [
    "Privacy Policy",
    "Политика конфиденциальности",
  ],
  [
    "Legal Notices",
    "Юридические уведомления",
  ],
  [
    "Unable to load webpage.",
    "Не удалось загрузить веб-страницу.",
  ],
  [
    "OK",
    "ОК",
  ],
  [
    "Do you want to add the token named “%@”?",
    "Добавить токен "%@"?",
  ],
  [
    "No",
    "Нет",
  ],
  [
    "Yes",
    "Да",
  ],
  [
    "Invalid Key",
    "Недействительный ключ",
  ],
  [
    "The key '%@' is invalid.",
    "Ключ "%@" недействителен.",
  ],
  [
    "You must enter a key.",
    "Необходимо ввести ключ.",
  ],
  [
    "Try Again",
    "Повторить попытку",
  ],
  [
    "Cancel",
    "Отмена",
  ],
  [
    "Invalid Barcode",
    "Недопустимый штрихкод",
  ],
  [
    "The barcode '%@' is not a valid authentication token barcode.",
    "Штрихкод "%@" не является действительным токеном аутентификации.",
  ],
  [
    "Welcome",
    "Добро пожаловать",
  ],
  [
    "Welcome_label",
    "Google Authenticator обеспечивает дополнительную безопасность аккаунта с помощью сгенерированных кодов.

Чтобы использовать Google Authenticator, необходимо добавить аккаунт, нажав:",
  ],
  [
    "Legal Information",
    "Юридическая информация",
  ],
  [
    "Authenticator",
    "Authenticator",
  ],
  [
    "Google Authenticator",
    "Google Authenticator",
  ],
  [
    "Add Token",
    "Добавить токен",
  ],
  [
    "Time Based",
    "По времени",
  ],
  [
    "Counter Based",
    "По счетчику",
  ],
  [
    "Account:",
    "Аккаунт:",
  ],
  [
    "user@example.com",
    "user@example.com",
  ],
  [
    "Key:",
    "Ключ:",
  ],
  [
    "Enter your key",
    "Введите ключ",
  ],
  [
    "Scan Barcode",
    "Сканировать штрихкод",
  ],
  [
    "Integrity Check Value",
    "Integrity Check Value",
  ],
]
`;

exports[`Localizations files helpers should read pass.strings file 2`] = `
[
  [
    "Terms of Service",
    "服务条款",
  ],
  [
    "Privacy Policy",
    "隐私权政策",
  ],
  [
    "Legal Notices",
    "法律声明",
  ],
  [
    "Unable to load webpage.",
    "无法加载网页。",
  ],
  [
    "OK",
    "确定",
  ],
  [
    "Do you want to add the token named “%@”?",
    "您想添加名为“%@”的令牌吗？",
  ],
  [
    "No",
    "否",
  ],
  [
    "Yes",
    "是",
  ],
  [
    "Invalid Key",
    "密钥无效",
  ],
  [
    "The key '%@' is invalid.",
    "密钥“%@”无效。",
  ],
  [
    "You must enter a key.",
    "您必须输入一个密钥。",
  ],
  [
    "Try Again",
    "重试",
  ],
  [
    "Cancel",
    "取消",
  ],
  [
    "Invalid Barcode",
    "条形码无效",
  ],
  [
    "The barcode '%@' is not a valid authentication token barcode.",
    "条形码“%@”不是有效的身份验证令牌条形码。",
  ],
  [
    "Welcome",
    "欢迎",
  ],
  [
    "Welcome_label",
    "Google 身份验证器可以生成验证码，从而增强登录时的安全性。

要使用 Google 身份验证器，必须先添加一个帐户，方法是按：",
  ],
  [
    "Legal Information",
    "法律信息",
  ],
  [
    "Authenticator",
    "身份验证器",
  ],
  [
    "Google Authenticator",
    "Google 身份验证器",
  ],
  [
    "Add Token",
    "添加令牌",
  ],
  [
    "Time Based",
    "基于时间",
  ],
  [
    "Counter Based",
    "基于计数器",
  ],
  [
    "Account:",
    "帐户：",
  ],
  [
    "user@example.com",
    "user@example.com",
  ],
  [
    "Key:",
    "密钥：",
  ],
  [
    "Enter your key",
    "输入您的密钥",
  ],
  [
    "Scan Barcode",
    "扫描条形码",
  ],
  [
    "Integrity Check Value",
    "Integrity Check Value",
  ],
]
`;
